### ISSUE TITLE
twist parametrization terminology / strange results when using pt2pl in filterreg

### ISSUE THREAD

**Issue by cmatija** [neka-nat/probreg#25]
Hi @neka-nat 

I have a confusion in your filterreg code when using the pt2pl mode I was hoping you could help me out with. In the function 'twist_trans' from se3_op.py, you write that it converts from twist representation to transformation matrix. According to various sources I found, this is typically done using e.g. rodrigues formula for the rotation matrix and a similar formula for the translation part as shown e.g. [here](https://robotics.stackexchange.com/questions/6599/how-to-get-the-44-matrix-from-the-twist-using-product-of-exponentials-in-robot). Now, while you use rodrigues formula for the rotation part, you simply return tw[3:] as the translation part.

On the other hand, looking at the system of equations you set up in point_to_plane.cc, it seems to me that you're following the technique described in [here](https://www.comp.nus.edu.sg/~lowkl/publications/lowk_point-to-plane_icp_techrep.pdf), in which case, however, the rotation matrix should be obtained from the alpha/beta/gamma using eq.5 in that document, and the representation is not actually a twist representation but instead euler angles?

Bottom line, I can't reconcile my understanding of parametrizing transformations with your code in the above 2 mentioned functions, and I'm asking because I sometimes get very strange results when using pt2pl mode in filterreg. Now, if I'm wrote with what I wrong above, could you please give me some pointers where I can additionally read up on this? I would really appreciate this a lot :)

Kind regards and thanks again for the great work!

**Comment by neka-nat** [comment]
Hi @cmatija ,

Thank you for the some suggestions.
In my opinion, there are two methods when converting from a twist to a transformation matrix.
One is *a linear approximation* and the other is *Rodrigues' rotation formula*.

I think your first opinion is correct. In the function I made, the rotation is computed using Rodrigues' formula, but the translation is a linear approximation. So, as you say, I think it's better to make the translation also Rodrigues' formula.

In the second opinion, eq.5 uses the linear approximation. However, to improve accuracy I have replaced it with Rodrigues' formula.

Finally, regarding the results of pt2pl, in general, pt2pl doesn't work well for objects with many curved surfaces. And I also found it difficult to set the sigma2 parameter when using the pt2pl filterreg.
If a stable solution is not obtained when using ICP's pt2pl, it is considered to be difficult even for filterreg.

**Comment by neka-nat** [comment]
I also had a question that came up.
The definition of twist described in [your link](https://robotics.stackexchange.com/questions/6599/how-to-get-the-44-matrix-from-the-twist-using-product-of-exponentials-in-robot) seems to relate to a robot joint, so is that the formula that can apply to the twist calculation in this case?

**Comment by cmatija** [comment]
I would think so. If you for instance look at the first chapter of 'An Invitation to Computer Vision' by Ma et al. you will find the same formula. 

That being said, I have also checked out the original code (which you probably did also) and they do it the same way, i.e. use Rodrigues formula for R but leave t untouched, so I guess I'm still missing something here.

**Comment by neka-nat** [comment]
Thank you for the good reference!
Is this the original code you're referring to?
https://github.com/sunglok/3dv_tutorial
I'll check it out as well.

**Comment by cmatija** [comment]
No I mean the original c++ filterreg code by the authors
